documentation/dataplane/dash-flow-api.md

```diff
@@ -0,0 +1,649 @@
+  
+    sai_flow_entry_t *flow_entry;
+  
+    uint32_t attr_count; 
+  
+    sai_attribute_t *attr_list; 
+
+} sai_flow_bulk_get_session_event_data_t;
+
+/**
+ * @brief dash flow get bulk session notification
+ *
+ * Passed as a parameter into sai_initialize_switch()
+ *
+ * @count data[count]
+ *
+ * @param[in] count Number of notifications
+ * @param[in] data Array of flow bulk get session events
+ */
+typedef void (*sai_flow_bulk_get_session_event_notification_fn)(
+        _In_ uint32_t count,
+        _In_ const sai_flow_bulk_get_session_event_data_t *flow_bulk_get_session_event_data);
+```
+
+### Protobuf-based flow programming
+
+In addition to the flow state attributes, the flow state can be represented using protobuf for high-efficiency. The attribute of the flow entry is SAI_FLOW_ENTRY_ATTR_FLOW_DATA_PB.
+
+Although the content of both attributes and protobuf may be identical, their applications differ. Attributes enable incremental updates to individual properties, whereas protobuf necessitates a complete update.
+
+```protobuf
+syntax = "proto3";
+
+message MacAddress {
+  bytes address = 1 [(validate.rules).bytes.len = 6];  // MAC address bytes
+}
+
+message IpAddress {
+  enum AddressType {
+    IPV4 = 0;
+    IPV6 = 1;
+  }
+
+  AddressType type = 1;  // IP address type (IPv4 or IPv6)
+  bytes address = 2;  // IP address bytes
+}
+
+message SaiDashFlowKey {
+  uint64 eni_mac = 1;  // ENI MAC address
+  IpAddress src_ip = 2;  // Source IP address
+  IpAddress dst_ip = 3;  // Destination IP address
+  uint32 src_port = 4;  // Source port
+  uint32 dst_port = 5;  // Destination port
+}
+
+message SaiDashFlowState {
+  uint32 version = 1;  // SAI_FLOW_ENTRY_ATTR_VERSION
+  uint32 dash_flow_action = 2;  // SAI_FLOW_ENTRY_ATTR_DASH_FLOW_ACTION
+  uint32 meter_class = 3;  // SAI_FLOW_ENTRY_ATTR_METER_CLASS
+  bool is_bidirectional_flow = 4;  // SAI_FLOW_ENTRY_ATTR_IS_BIDIRECTIONAL_FLOW
+  uint32 underlay_vni = 5;  // SAI_FLOW_ENTRY_ATTR_UNDERLAY_VNI
+  IpAddress underlay_sip = 6;  // Underlay source IP address
+  IpAddress underlay_dip = 7;  // Underlay destination IP address
+  MacAddress underlay_smac = 8;  // Underlay source MAC address
+  MacAddress underlay_dmac = 9;  // Underlay destination MAC address
+  uint32 underlay2_vni = 10;  // SAI_FLOW_ENTRY_ATTR_UNDERLAY2_VNI
+  IpAddress underlay2_sip = 11;  // Underlay2 source IP address
+  IpAddress underlay2_dip = 12;  // Underlay2 destination IP address
+  MacAddress underlay2_smac = 13;  // Underlay2 source MAC address
+  MacAddress underlay2_dmac = 14;  // Underlay2 destination MAC address
+  MacAddress dst_mac = 15;  // Destination MAC address
+  IpAddress sip = 16;  // Source IP address
+  IpAddress dip = 17;  // Destination IP address
+  bytes sip_mask = 18;  // Source IP mask
+  bytes dip_mask = 19;  // Destination IP mask
+}
+
+message SaiDashFlowEntry {
+  SaiDashFlowKey flow_key = 1;
+  SaiDashFlowKey reverse_flow_key = 2;
+  SaiDashFlowState flow_state = 3;
+}
```

Comment:
What about the flow state for the reverse direction?

Reply:
The reverse_flow_key works as the pointer of the reverse flow. User should use the pointer to get the real reverse flow state.